### PR TITLE
Create Allowed External IPs PSP

### PR DESCRIPTION
**Pull request by ZeroMagic**
Ref: https://github.com/Azure/AKS/issues/2012

### CODE REVIEW

File: Policies/Kubernetes/allowed-external-ips/rego/src.rego [hunk 1]
```diff
@@ -0,0 +1,11 @@
+package k8sazureexternalips
+
+violation[{"msg": msg}] {
+  input.review.kind.kind == "Service"
+  input.review.kind.group == ""
+  allowedIPs := {ip | ip := input.parameters.allowedIPs[_]}
+  externalIPs := {ip | ip := input.review.object.spec.externalIPs[_]}
+  forbiddenIPs := externalIPs - allowedIPs
+  count(forbiddenIPs) > 0
+  msg := sprintf("service has forbidden external IPs: %v", [forbiddenIPs])
```
Comment (EmandM) [inline Policies/Kubernetes/allowed-external-ips/rego/src.rego:10]: Can you change the message to include the list of allowed IPs? `"Service has forbidden external IPs: %v, Allowed IPs: %v"`

**Review by EmandM — APPROVED** [review]

File: Policies/Kubernetes/allowed-external-ips/rego/src.rego (hunk 1 above)
Comment (ZeroMagic) [reply Policies/Kubernetes/allowed-external-ips/rego/src.rego:10]: I am not sure it is by design that the allowed IPs are not shown. Let's check it with @ritazh.

File: Policies/Kubernetes/allowed-external-ips/examples-good/azurepolicy.assignment.parameters.json [hunk 2]
```diff
@@ -0,0 +1,6 @@
+{
+  "parameters": {
+    "excludedNamespaces": { "value": [ "kube-system", "azure-arc" ] },
+    "allowedExternalIPs": { "value": [ "203.0.113.0"] }
```
Comment (sozercan) [inline Policies/Kubernetes/allowed-external-ips/examples-good/azurepolicy.assignment.parameters.json:4]: do we need a default one?

File: Policies/Kubernetes/allowed-external-ips/rego/src.rego (hunk 1 above)
Comment (ritazh) [reply Policies/Kubernetes/allowed-external-ips/rego/src.rego:10]: I think for now we should follow what upstream https://github.com/open-policy-agent/gatekeeper-library/blob/master/library/general/externalip/template.yaml#L32 is doing for this policy. I can see cases where displaying the allowed ip might not be a good idea. Let's revisit this if a user asks for the allowed IP to be returned.

File: Policies/Kubernetes/allowed-external-ips/examples-good/azurepolicy.assignment.parameters.json (hunk 2 above)
Comment (ZeroMagic) [reply Policies/Kubernetes/allowed-external-ips/examples-good/azurepolicy.assignment.parameters.json:4]: I don't think so. We don't know the IPs that users can use.

Comment (sozercan) [reply Policies/Kubernetes/allowed-external-ips/examples-good/azurepolicy.assignment.parameters.json:4]: can we leave this blank then?

Comment (ZeroMagic) [reply Policies/Kubernetes/allowed-external-ips/examples-good/azurepolicy.assignment.parameters.json:4]: Yep. I referred to other policies. Some of them leave the blank which means that all the values are not allowed. 
Here the line is just an example. The real value in our policy is located at `azurepolicy.json`
```
"allowedExternalIPs": {
        "type": "Array",
        "metadata": {
          "displayName": "Allowed External IPs",
          "description": "List of External IPs that services are allowed to use. Empty array means all external IPs are disallowed."
        },
        "defaultValue": []
      }
```
```
"allowedExternalIPs": "[parameters('allowedExternalIPs')]"
```
Now the default value is blank.

Comment (sozercan) [reply Policies/Kubernetes/allowed-external-ips/examples-good/azurepolicy.assignment.parameters.json:4]: ah ok i see, looks like this file is just an example.

File: Policies/Kubernetes/allowed-external-ips/azurepolicy.json [hunk 3]
```diff
@@ -0,0 +1,65 @@
+{
+  "properties": {
+    "displayName": "Ensure only allowed External IPs are used in Kubernetes Cluster",
+    "policyType": "Custom",
+    "mode": "Microsoft.Kubernetes.Data",
+    "description": "This policy ensures only allowed external ips are used in a Kubernetes cluster. For instructions on using this policy, please visit https://aka.ms/kubepolicydoc",
+    "metadata": {
+      "version": "1.0.0-preview",
+      "category": "Kubernetes PSP",
```
Comment (sozercan) [inline Policies/Kubernetes/allowed-external-ips/azurepolicy.json:9]: this is not a PSP

File: Policies/Kubernetes/allowed-external-ips/azurepolicy.parameters.json [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    "metadata": {
+      "displayName": "Namespace exclusions",
+      "description": "List of Kubernetes namespaces to exclude from policy evaluation. Providing a value for this parameter is optional."
+    },
+    "defaultValue": ["kube-system", "gatekeeper-system", "azure-arc"]
+  },
+  "allowedExternalIPs": {
+    "type": "Array",
+    "metadata": {
+      "displayName": "Allowed Exeternal IPs",
```
Comment (sozercan) [inline Policies/Kubernetes/allowed-external-ips/azurepolicy.parameters.json:26]:
```suggestion
      "displayName": "Allowed External IPs",
```

File: Policies/Kubernetes/allowed-external-ips/azurepolicy.json (hunk 3 above)
Comment (ZeroMagic) [reply Policies/Kubernetes/allowed-external-ips/azurepolicy.json]: Good catch. It is a service, not pod. Fixed.

File: Policies/Kubernetes/allowed-external-ips/azurepolicy.parameters.json (hunk 4 above)
Comment (ZeroMagic) [reply Policies/Kubernetes/allowed-external-ips/azurepolicy.parameters.json]: done

**Pull request merged**